Python/ceval.c

```diff
@@ -863,6 +863,75 @@ static const binaryfunc binary_ops[] = {
     [NB_INPLACE_XOR] = PyNumber_InPlaceXor,
 };
 
+#define NEXT_OP_STORES(O)                                  \
+    ((_Py_OPCODE(*next_instr) == STORE_FAST ||             \
+      _Py_OPCODE(*next_instr) == STORE_FAST__LOAD_FAST) && \
+     GETLOCAL(_Py_OPARG(*next_instr)) == (O))              \
+
+#define BINARY_OP_FAST_INT(OP)                                                 \
+    do {                                                                       \
+        PyObject *lhs = SECOND();                                              \
+        PyObject *rhs = TOP();                                                 \
+        DEOPT_IF(!PyLong_CheckExact(lhs), BINARY_OP);                          \
+        DEOPT_IF(!PyLong_CheckExact(rhs), BINARY_OP);                          \
+        DEOPT_IF(1 < Py_ABS(Py_SIZE(lhs)), BINARY_OP);                         \
+        DEOPT_IF(1 < Py_ABS(Py_SIZE(rhs)), BINARY_OP);                         \
+        STAT_INC(BINARY_OP, hit);                                              \
+        stwodigits l = Py_SIZE(lhs) * ((PyLongObject *)lhs)->ob_digit[0];      \
+        stwodigits r = Py_SIZE(rhs) * ((PyLongObject *)rhs)->ob_digit[0];      \
+        stwodigits i = l OP r;                                                 \
+        Py_DECREF(rhs);                                                        \
+        STACK_SHRINK(1);                                                       \
+        if (-_PY_NSMALLNEGINTS <= i && i < _PY_NSMALLPOSINTS) {                \
+            Py_DECREF(lhs);                                                    \
+            PyLongObject *res = &_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS + i];    \
+            Py_INCREF(res);                                                    \
+            SET_TOP((PyObject *)res);                                          \
+            DISPATCH();                                                        \
+        }                                                                      \
+        bool inplace = NEXT_OP_STORES(lhs);                                    \
+        if (Py_ABS(i) < PyLong_BASE && Py_REFCNT(lhs) == inplace + 1) {        \
+            assert(lhs < (PyObject *)&_PyLong_SMALL_INTS[0] ||                 \
+                   lhs >= (PyObject *)&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS +  \
+                                                          _PY_NSMALLPOSINTS]); \
+            ((PyLongObject *)lhs)->ob_digit[0] = Py_ABS(i);                    \
```

Comment:
I make it 10 or 11 tests to get to here. This seems excessive for a specialization.
I think the real problem is that our `int` implementation is a bit clumsy for ints that fit into a single machine word.
I'd like to address that problem first, rather than make this unmaintainable.

How to you plan to test all the possible paths here?